lib/models/addon.js

```diff
@@ -371,16 +391,76 @@ let addonProto = {
   },
 
   /**
-    Discovers all child addons of this addon and stores their names and
-    package.json contents in this.addonPackages as key-value pairs
-
-    @private
-    @method discoverAddons
+   * Discovers all child addons of this addon and an AddonInfo about
+   * each addon in this.addonPackages (keyed on addon name).
+   *
+   * Child addons include those from 'dependencies' (not devDependencies)
+   * and in-repo addons
+   *
+   * Any packageInfos that we find that are marked as not valid are excluded.
+   *
+   * @private
+   * @method discoverAddons
    */
   discoverAddons() {
-    let addonsList = this.addonDiscovery.discoverChildAddons(this);
+    let pkgInfo = this._packageInfoCache.getEntry(this.root);
+
+    let addonPackageList = [];
+
+    // There are cases where an addon can be created in memory and not actually
+    // have a root entry (or have one that is not actually pointing to a directory,
+    // like 'foo' in some of the tests. We don't want to crash, but want to let
+    // things continue even though they're empty.
+    if (pkgInfo) {
+
+      // The 'dependencyPackages' object contains ALL dependencies, not just those
+      // that are addons. We want only the addons.
+      if (pkgInfo.dependencyPackages) {
+        Object.keys(pkgInfo.dependencyPackages).forEach(name => {
```

Comment:
can we use `filter` here?

Reply:
Not really, because the list contains names, and the result is additions of their corresponding values to the list--we're not just filtering the names. If we could have used Object.values() it would have been okay, but that's not in ES5.

Reply:
wonder if we could use [`entries`](https://node.green/#ES2015-built-in-extensions-Array-prototype-methods-Array-prototype-entries)

Reply:
Actually, it doesn't work. pgkInfo.dependencyPackages is a plain JS object (i.e., a map), so Array.entries won't work directly. If you use Object.keys() on it first, then you have an array of names, which means you still need to get the value from the dependencyPackages property. We really need Object.values(), which does exactly what we want.

Reply:
the following may work, although I'm not sure its worth changing.

```
Object.entries(pkgInfo.dependencyPackages).forEach(([ name, dependencyInfo ]) => console.log('name', name, 'depInfo', dependencyInfo))
```